R/Rules-methods.R

```diff
@@ -284,77 +284,59 @@ setMethod(
-              nPatients <- sapply(dltSplit, length)
-              nDLTs <- sapply(dltSplit, sum)
+#' @describeIn nextBest find the next best dose based on the 3+3 method.
+#'
+#' @aliases nextBest-NextBestThreePlusThree
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestThreePlusThree.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestThreePlusThree",
+    doselimit = "missing",
+    samples = "missing",
+    model = "missing",
+    data = "Data"
+  ),
+  definition = function(nextBest, doselimit, samples, model, data, ...) {
 
-              ## what was the last dose level tested?
-              lastLevel <- tail(data@xLevel, 1)
+    # The last dose level tested (not necessarily the maximum one).
+    last_level <- tail(data@xLevel, 1L)
 
-              ## if there are less than 1/3 DLTs at that level
-              if(nDLTs[lastLevel]/nPatients[lastLevel] < 1/3)
-              {
-                  ## we could escalate, unless this is the highest
-                  ## level or the higher level was tried already
-                  ## (in which case it was not safe)
-                  if((lastLevel == length(data@doseGrid)) ||
-                     (nPatients[lastLevel+1] > 0))
-                  {
-                      nextLevel <- lastLevel
-                  } else {
-                      nextLevel <- lastLevel + 1
-                  }
-              } else if(nDLTs[lastLevel]/nPatients[lastLevel] > 1/3) {
-                  ## rate here is too high, therefore deescalate
-                  nextLevel <- lastLevel - 1
-              } else {
-                  ## otherwise: rate is 1/3 == 2/6,
-                  ## then it depends on the number of patients:
-                  ## if more than 3, then deescalate, otherwise stay.
-                  nextLevel <-
-                      if(nPatients[lastLevel] > 3)
-                          lastLevel - 1
-                      else
-                          lastLevel
-              }
+    # Get number of patients per grid's dose and DLT rate at the last level.
+    nPatients <- table(factor(data@x, levels = data@doseGrid))
+    nDLTs_last_level <- sum(data@y[data@xLevel == last_level])
+    DLT_rate_last_level <- nDLTs_last_level / nPatients[last_level]
 
-              ## do we stop here? only if we have no MTD
-              ## or the next level has been tried enough (more than
-              ## three patients already)
-              stopHere <-
-                  if(nextLevel == 0)
-                  {
-                      TRUE
-                  } else {
-                      nPatients[nextLevel] > 3
-                  }
+    next_level <- last_level + if (DLT_rate_last_level < 1 / 3) {
+      # Escalate it, unless this is the highest level or the higher level
+      # already tried (and it was not safe).
+      ifelse((last_level == data@nGrid) || (nPatients[last_level + 1] > 0), 0, 1)
```

Comment:
Hi @danielinteractive , I have one doubt to the comment that is in the brackets in the line 317, that is I think it if not true for the code in the line 318, as it does not check whether the higher level that was already tried is safe or not.

Reply:
Hi @wwojciech well implicitly we know that it was not safe - because only then we would ever go down again

Reply:
Tx @danielinteractive , so it means that we should assume that such data is not possible?
```
data <- Data(
  x = c(25, 25, 100, 75, 75, 75, 75),
  y = c(0L, 0L, 0L, 0L, 0L, 0L, 1L),
  doseGrid = c(25, 50, 75, 100, 125)
)
```
Because, here the last dose it `75` and the higher `100` was already tried before and it did not have any DLT...

Reply:
such data won't be possible I think. feel free to modify the comment though if you think too many assumptions behind it

Reply:
I would not change the algorithm though.